### ISSUE TITLE
Propose adding an option to deploy by copying then moving instead of symlinking

### ISSUE THREAD

**Issue by nathanstretch** [capistrano/capistrano#985]
The reason for this proposal is that Apache (and likely other web servers) cache symlinks for a period of time, so when the 'current' symlink is updated to a new release, the old release is still served for a period of time (up to a minute or so). More details here: https://stackoverflow.com/questions/18450076/capistrano-symlinks-being-cached/19630266#19630266.

This can cause a number of problems, particularly when combined with caching.

One solution is to restart apache on each deploy, but that's not ideal. Alternatively, instead of a symlink, the release could actually be moved to current rather than symlinked. The process would look something like:

```
1. Generate new directory under releases, exactly as is done now.
2. Copy the new release directory to releases/new
3. mv current releases/old && mv releases/new current
4. rm -rf releases/old
```

The rollback process would work the same, except that it would use the previous release directory for step two instead of generating a new one.

I don't feel nearly knowledgeable about the code base to write a pull request for this, since there may well be edge cases or other problems I know nothing about, but on the surface this seems fairly straightforward.  Any possibility of adding an option like this?  If not, is there any reason I couldn't (or shouldn't) add a replacement for :symlink:release in my deploy.rb to have it do this?

**Comment by DonGiulio** [comment]
I'm using Capistrano 3.4.0 to deploy our RoR4 application on a shared SMB mounted disk on Azure File Share.

The problem is that this SMB disk doesn't support symlinks, therefore I can't deploy using Capistrano as it is now. 

The problem occurs when trying to create the symlinks for the configuration files, and even if I fix these, will occur also when creating the `current` symlink.  

I think the ability to copy/move files instead of symlinking them would solve my issue. 

so +1 on this change

**Comment by leehambley** [comment]
Thanks for weighing in @DonGiulio - this is another vote in favour of the changes you suggested, however it will likely be a part of a migration (maybe I have time in September, I have some time off work) to extract the Git, and Subversion strategies, and at the same time extract the symlink release strategy. It's pretty brutal core surgery, but it's high time I stopped talking, and started coding!

**Comment by DonGiulio** [comment]
I'm writing some code for it, I'm not sure if it's what you have in mind, but I will make a PR, so you can see it.

**Comment by nathanstretch** [comment]
FYI, with the switch from APC to Opcache in php, there's more reason to use atomic move rather than symlinking for deployment, as described here: http://codinghobo.com/opcache-and-symlink-based-deployments/

It is important to make sure it's still an instantaneous, atomic operation though (as a move swap should be), as opposed to say a copy, which would take time and result in inconsistency/downtime.  Of course  you could optionally make a copy of the release before moving it into place.

**Comment by joshribakoff** [comment]
I do not use Capistrano, I use my own script(s) that deploy using symlinks. I've done this for years using both nginx & apache without cache issues. However today I ran into the cache issues on both nginx & apache.

Its something that has changed in PHP, between 5.0.x and 5.6.x there has been some change to the PHP SAPI / php-fpm, where it caches the locations of files to be executed. Setting `realpath_cache_size` to 0 and setting `realpath_cache_ttl` to 1 fix the cache issue for both nginx & Apache.

Its not specific to Apache at all. This is a PHP.ini setting that is "new", which affects both mod_apache & php-fpm, regardless of if you use apache, nginx, or some other server.

**Comment by nathanstretch** [comment]
Josh, you're right that there is a new, similar issue with php 5.6. That
issue is the use of zend opcache, which does not realize a change has been
made when a symlink is updated.

Fortunately, since you're using fpm, all your processes share a cache, so
if you clear it (the one used by fpm, not the cli) after updating the
symlink, it should solve your problem.

--
Nathan

**Comment by joshribakoff** [comment]
No.. to be clear, we observed the caching of symlinks' targets with both Apache mod_php & nginx w/ php-fpm, with and without opcache enabled. Opcache or nginx may be caching the bytecode of the resulting PHP file after the symlink target is looked up, but PHP itself is caching the targets of symlinks, even with opcache disabled. Setting `realpath_cache_size` to 0 fixed the problems.

However you are correct @nathanstretch that even if `realpath_cache_size` is set to 0, opcache may be separately caching the bytecode for the old PHP file even though PHP sees the new symlink target.

If you want your updates to take place, you need to be aware of all of these things that are involved.

**Comment by nathanstretch** [comment]
Totally.  I just assumed since you mentioned switching to 5.6 it was likely
opcache that was getting you.  But yes, there are a number of gotchas;
ultimately moving away from symlink-based deployment is probably wise.


--
Nathan Stretch - CEO

www.searchtempest.com - All the Classifieds. One Search.
www.autotempest.com - Everyone's ideal car is out there. Find yours.

**Comment by jrean** [comment]
@nathanstretch I ended here, I'm encountering the same kind of issue. I'm using my own scripts to deploy and when I point my symlink to my new build, the link is not updated... I just tried what you both suggested but it doesn't work.

- Did you come with a working solution? (I'm on PHP7)

>moving away from symlink-based deployment is probably wise

What do you mean?

**Comment by jrean** [comment]
If it can help someone, this tool is just awesome and will do what has to be done...
`http://gordalina.github.io/cachetool/`
